### ISSUE TITLE
Login Method

### ISSUE THREAD

**Issue by krulci** [fa0311/DMMGamePlayerFastLauncher#43]
I was not able to login.
But I checked their API.
So if you send the cookie and all other headers correctly to https://apidgp-gameplayer.games.dmm.com/v5/loginurl
You will get a json in the response code.
The json will contain an url with the correct login token.
If you again send the cookie correctly with all other headers it will have two 302 redirect.
The one where the URI contains /service/login/token/ is the one with login_secure_id and login_session_id.
Also, I think the reason that you mentioned DMMGamePlayer must re-login everytime this app is used is because you never write the cookie the server response provide to the db.
Basically, server will return cookies with "set-cookie" key.
These key and value should be used to update the Cookies db under Network folder.
This will ensure every auto login uses the last cookie retrieve from the server.

**Comment by krulci** [comment]
I went an implemented this.
So there are basically 3 API involved in the auto login process with a few redirection.
```
https://apidgp-gameplayer.games.dmm.com/v5/loginurl
```
This API is responsible for generating a tokenized login url.
I gave it the proper headers; however, I think the headers don't matter with this API.
You can just open this in any browser and get a working tokenized login url with result code 100.
```
https://accounts.dmm.com/service/login/password/=/path={Token}?device=games-player
```
This API is responsible for automatic password authentication.
This API if handle properly will result in a 302 result code with URI to the next API.
You will need to add all cookies related to the host of this API from the cookie database.
You get three login cookies back from properly handling this API:
```
Set-Cookie: ckcy=2; path=/; expires={removed_for_privacy}; domain=dmm.com; samesite=none; secure
Set-Cookie: cklg=ja; path=/; expires={removed_for_privacy}; domain=dmm.com; samesite=none; secure
Set-Cookie: connect.sid={removed_for_privacy}; Path=/; HttpOnly; Secure; SameSite=None
```
You then update the cookie database with these if it exists.
Based on my observation, connect.sid is the only one that does not exist.
```
https://accounts.dmm.com/service/login/token/=/path={Token}/is_app=false
```
This API is responsible for the actual authentication.
This API if handle properly will result in a 302 result code with URI to the next API.
You will need to add all cookies related to the host of this API from the cookie database.
This API required not only cookies read from the cookie db but also "Set-Cookie" from the previous API.
Similar to the previous API, this API if handle properly will return 12 "Set-Cookie" key value pairs.
These include "login_secure_id" and "login_session_id".
You then update the cookie database with these if it exists.
```
https://www.dmm.com/my/-/redirect/=/rurl={Token}
```
This API indicates login success.
This API if handle properly will result in a 200 result code.
You will need to add all cookies related to the host of this API from the cookie database.
This API required not only cookies read from the cookie db but also "Set-Cookie" from the previous API.
Note, this API has a different Host URI comparing to the previous two API which indicates the cookies read from db is different.
Similar to the previous API, this API if handle properly will return 3 "Set-Cookie" key value pairs.
You then update the cookie database with these if it exists.
```
https://apidgp-gameplayer.games.dmm.com/v5/launch/cl
```
This API returns the execution argument for the requested product.
This API if handle properly will result in a 200 result code.
This API requires 3 cookies.
```
age_check_done=0
login_secure_id={secure_id}
login_session_id={session_id}
```
Based on my observation, both age_check_done and login_session_id does not get saved to the cookie database from the previous APIs.
I just assign age_check_done as 0 and pass login_session_id from the /servicee/login/token/ API.

These are all it takes for auto login as long as dmmgameplayer5 is installed and username and password is saved and the actually client is not signed out.

I also check their non auto login API.
Which is the:
```
https://accounts.dmm.com/service/login/password/authenticate
```
But it seems like it is using Google's reCaptcha to generate a specific token.
Also the tokenized login url's path parameter seems to instead be generated by the client.
Unlike the auto login method where you just get a tokenized login url by requesting a specific API.
There is also another random token which seems to be gerated by the client.
I judge the path parameter and the non Google reCaptcha token to be random because it is not in the response header or content of the Google reCaptcha API and there is no other API being accessed before the autenticate API.
These three along with unencrypted username and password are included in the content of the request.
I tried to dissambled the client but the decompiled javascrirpt code is obfuscated and also a big nasty chunck so I just gave up on the login process.

**Comment by fa0311** [comment]
Thank you for sending your Issue!

> I was not able to login.

Hmmm, I can log in in my environment.
This problem has already been reported, but unfortunately I can't reproduce it in my environment #34

> Also, I think the reason that you mentioned DMMGamePlayer must re-login everytime this app is used is because you never write the cookie the server response provide to the db.

This will be implemented.

> This API is responsible for generating a tokenized login url.

This software does not use this endpoint
Retrieved from #auto-login on https://www.dmm.com/
[source](https://github.com/fa0311/DMMGamePlayerFastLauncher/blob/master/DMMGamePlayerFastLauncher.py#L101-L108)
However, we are considering using this endpoint.

The request library used by this software automatically handles redirection and cookie management
https://requests.readthedocs.io/en/latest/

**Comment by fa0311** [comment]
We have released [v4.5.0](https://github.com/fa0311/DMMGamePlayerFastLauncher/releases/tag/v4.5.0) which fixes this.
Please check.

**Comment by krulci** [comment]
I just tried v4.5.1.
Although I wasn't able to start DFL while fiddler is running to inspect the request and response, but the new version works flawless without login out the genuine DMMGamePlayer.
It seems like v4.5.1 loads all the cookie first and let request library handles redirection and cookie automatically, finally saving the cookies to database after all the request has completed.
I handle cookie and redirection manually because HttpClient in .Net faultly set duplicated cookies upon redirection and "set-cookie".
Might just be because the library has an issue or I wasn't using it correctly.
